File: R/alexnet-dropout.R
dropout <- function(x, p = 0.5, training = TRUE) {
  if (!training || p == 0) {
    return(x)
  }

  mask <- rbinom(length(x), size = 1, prob = 1 - p)
  mask <- array(mask, dim = dim(x))
  (x * mask) / (1 - p)
}

File: pytorch-mps/alexnet-dropout.py
import torch


def dropout(x: torch.Tensor, p: float = 0.5, training: bool = True, device=None) -> torch.Tensor:
    if device is None:
        device = "mps" if torch.backends.mps.is_available() else "cpu"
    x = x.to(device)
    if not training or p == 0:
        return x

    mask = torch.bernoulli(torch.full_like(x, 1 - p))
    return (x * mask) / (1 - p)
